controllers/installation_controller.go

```diff
@@ -369,8 +387,14 @@ func (r *InstallationReconciler) ReconcileK0sVersion(ctx context.Context, in *v1
 	// of the plan id is deprecated in favour of the annotation.
 	annotation := plan.Annotations[InstallationNameAnnotation]
 	if annotation == in.Name || plan.Spec.ID == in.Name {
-		r.SetStateBasedOnPlan(in, plan)
-		return nil
+		// there are two plans needed to be run in sequence for airgap upgrades. the first one is
+		// the one that copies the artifacts to the nodes and the second one is the one that
+		// actually upgrades the k0s version. we need to make sure that this is the second plan
+		// before setting the installation state to the plan state.
+		if isAutopilotUpgradeToVersion(&plan, desiredVersion) {
```

Comment:
this was needed otherwise an upgrade was never attempted as it would set the status to kubernetesinstalled based on the distribute airgap artifacts plan. the rest is just moving some code around.